### DISCUSSION TITLE
Implement multiple JWT lifetimes for different routes

### DISCUSSION THREAD

**Discussion by jakemanger** [fastapi-users/fastapi-users#775] [Q&A] [answered]
Thanks a lot for this amazing library - it's making my life a lot easier.

Is it possible to have multiple JWT lifetimes that can only be used for certain routes?
My use case is where I have the user use a JWT token that doesn't expire for a long time to access one route, but for more secure things (e.g. changing password) I want the user to use a JWT token that was emailed to them that only lasts a small amount of time.

If so, how would I implement this with the current SQLAlchemy example found at https://fastapi-users.github.io/fastapi-users/configuration/full-example/?

**Comment by frankie567** [comment, marked as answer]
Yes, you can start with the same basis. What you need then to do is the following:

1. Have two JWT authentication classes, one for "long" tokens and one for "slow" tokens.

```py
jwt_authentication_long = JWTAuthentication(
    secret="SECRET_FOR_LONG_JWT", lifetime_seconds=86400, tokenUrl="auth/jwt-long/login"
)
jwt_authentication_short = JWTAuthentication(
    secret="SECRET_FOR_SHORT_JWT", lifetime_seconds=3600, tokenUrl="auth/jwt-short/login"
)
```

2. Declare them both on the FastAPI Users class:

```py
fastapi_users = FastAPIUsers(
    get_user_manager,
    [jwt_authentication_long, jwt_authentication_short],
    User,
    UserCreate,
    UserUpdate,
    UserDB,
)
```

3. Wire both routers:

```py
app.include_router(
    fastapi_users.get_auth_router(jwt_authentication_long), prefix="/auth/jwt-long", tags=["auth"]
)
app.include_router(
    fastapi_users.get_auth_router(jwt_authentication_short), prefix="/auth/jwt-short", tags=["auth"]
)
```

4. The most important part is to add logic to determine which type of token to accept following the current route (see https://fastapi-users.github.io/fastapi-users/usage/current-user/#dynamically-enable-authentication-backends):

```py
async def get_enabled_backends(request: Request):
    """Return the enabled dependencies following custom logic."""
    if request.url.path == "/super-sensitive-route":  # Adapt it here following your own logic
        return [jwt_authentication_short]
    else:
        return [jwt_authentication_long, jwt_authentication_short]


current_active_user = fastapi_users.current_user(active=True, get_enabled_backends=get_enabled_backends)
```

**Reply by jakemanger** [reply]
Thank you so much. I'll give that a try. You've made a great project here and I am also thoroughly enjoying your new book!

**Reply by frankie567** [reply]
Glad you like them, much appreciated 🙏

**Reply by frankie567** [reply]
After a second thought, I forgot an important thing: be sure to have different secrets for each authentication classes; otherwise, they could be validated in both cases, which is obviously what you want to avoid. I've edited my first answer!

**Reply by jakemanger** [reply]
Great! Thanks for the change!

**Reply by jakemanger** [reply]
I got this working with one additional change needed:
> ```python
> app.include_router(
>     fastapi_users.get_auth_router(jwt_authentication_long), prefix="/auth/jwt-long", tags=["auth"]
> )
> app.include_router(
>     fastapi_users.get_auth_router(jwt_authentication_short), prefix="/auth/jwt-short", tags=["auth"], name='short_jwt'
> )
There was an error message complaining I had duplicate routers of the same name. Now that I have have another name there, it is working as expected.
